### ISSUE TITLE
Why my program takes 6 minutes to calculate gradient of a function first time but 0.3 seconds for second time?

### ISSUE THREAD

**Issue by Dream7-Kim** [google/jax#1399]
```python
import jax
from jax import jit
from jax import jacrev
import jax.numpy as np
import numpy as onp
import time

NUM_VAR = 10**4
NUM_ARG = 200

var_onp = onp.random.rand(NUM_VAR)
var = np.asarray(var_onp) # allocating the variables to GRAM

def func(arg):
    divider = 0 # denominator
    numerator = 0
    for i in range(NUM_ARG):
        temp = np.dot(arg[i], var)
        temp1 = np.sin(temp)
        temp2 = np.cos(temp)
        
        divid = np.add(temp1, temp2)
        divid = np.power(divid, 2)
        divid = np.sum(divid)

        numer = np.add(temp1, temp2)
        numer = np.sum(numer)
        numer = np.power(numer, 2)
        numerator = np.add(numer, numerator)

        divider = np.add(divider, divid)
    divider = np.power(divider, 1/2)

    return np.log(np.divide(numerator, divider))

res = jax.jit(jax.grad(func))

# arg = jax.random.normal(jax.random.PRNGKey(1), (NUM_ARG, 0))
arg = onp.random.rand(NUM_ARG)
real_args = np.asarray(arg)
print("... ... ...Start calculating Gradient... ... ...")
start = time.time()
print(res(real_args))
end = time.time()
print("First Execution time: ", end-start, " s ... ... ...")


for step in range(10):
    arg = onp.random.rand(NUM_ARG)
    real_args = np.asarray(arg)
    # print("... ... ...Start calculating Gradient... ... ...")
    start = time.time()
    res(real_args)
    end = time.time()
    print(str(step) + "th Execution time: ", end-start, " s ... ... ...")        

```

**Comment by Dream7-Kim** [comment]
It takes 6 minutes to calculate the gradient of a function for first time and 0.3 seconds for second and third time...
Why does this happen?
Because of the time to build auto-gradient graph or ?
Or because the definition of function includes lot of iterations?

Thank you for you reply and advice!!!

**Comment by mattjj** [comment]
Thanks for the question, and for the clear repro!

The differentiation step itself is fast; I added timing around [this line](https://github.com/google/jax/blob/762b602f3316f71f65b665f48dba465e96b2e392/jax/interpreters/xla.py#L381) and measured it took less than 1.6 seconds for JAX to do the differentiation and form the "jaxpr" data structure to be handed to XLA. I put timing around [this line](https://github.com/google/jax/blob/762b602f3316f71f65b665f48dba465e96b2e392/jax/interpreters/xla.py#L202) to find it took an additional 0.26 seconds to build an XLA computation out of that jaxpr data structure.

The slow part comes next: asking XLA to compile the computation. As you already guessed, because the function includes a lot of iterations with large `NUM_ARG`, that means a large-ish computation (the jaxpr has 7410 lines in it for a `NUM_ARG` value of 200). Because XLA goes to extreme effort to optimize the heck out of computations, its compilation time can scale super-linearly with the size of the computation graph we give it. (This behavior varies based on backend: often the XLA:CPU backend is the slowest to compile, and XLA:GPU is faster. Everything is getting faster all the time, though.)

Luckily we can do more than wait for the XLA team to make their amazing compiler run faster on all backends: we can make the program size not scale with `NUM_ARG` by using a control flow construct like [`lax.scan`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html#jax.lax.scan) which compiles down to a single node in the XLA graph.

Here's a new version using [`lax.scan`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html#jax.lax.scan):

```python
import jax
from jax import jit
from jax import jacrev
import jax.numpy as np
import numpy as onp
import time
from jax import lax

NUM_VAR = 10**4
NUM_ARG = 200

rng = onp.random.RandomState(0)  # NOTE: control RNG for reproducibility

var_onp = rng.rand(NUM_VAR)
var = np.asarray(var_onp) # allocating the variables to GRAM

def func(arg):
    divider = 0. # NOTE: I made these floats
    numerator = 0.

    def body_fun(carry, x):
        divider, numerator = carry
        temp = np.dot(x, var)
        temp1 = np.sin(temp)
        temp2 = np.cos(temp)

        divid = np.add(temp1, temp2)
        divid = np.power(divid, 2)
        divid = np.sum(divid)

        numer = np.add(temp1, temp2)
        numer = np.sum(numer)
        numer = np.power(numer, 2)
        numerator = np.add(numer, numerator)

        divider = np.add(divider, divid)

        new_carry = divider, numerator
        return new_carry, ()
    (divider, numerator), _ = lax.scan(body_fun, (divider, numerator), arg)
    divider = np.power(divider, 1/2)

    return np.log(np.divide(numerator, divider))

res = jax.jit(jax.grad(func))

# arg = jax.random.normal(jax.random.PRNGKey(1), (NUM_ARG, 0))
arg = rng.rand(NUM_ARG)
real_args = np.asarray(arg)
print("... ... ...Start calculating Gradient... ... ...")
start = time.time()
print(res(real_args))
end = time.time()
print("First Execution time: ", end-start, " s ... ... ...")


for step in range(10):
    arg = rng.rand(NUM_ARG)
    real_args = np.asarray(arg)
    # print("... ... ...Start calculating Gradient... ... ...")
    start = time.time()
    res(real_args).block_until_ready()  # NOTE: added!
    end = time.time()
    print(str(step) + "th Execution time: ", end-start, " s ... ... ...")
```

I quickly checked correctness by setting `NUM_ARG = 3` and comparing the two versions against each other numerically. I also added the `block_until_ready()` call because otherwise you'll measure [near-instant execution times thanks to asynchronous dispatch](https://jax.readthedocs.io/en/latest/async_dispatch.html).

**Comment by Dream7-Kim** [comment]
Thanks for you advice and help..

I solved the compile time problem by modifying the definition of the function, I have removed the iteration in the definition and it worked. The time is now 0.7 seconds with NUM_VAR=10**6.

The execution time of your code is very fast. But I cannot completely understand the usage of the lax.scan. I will read and read the description of lax.scan and try to understand it.

Now I have some other problems, hope you can help me with these problems.
My video card is NVIDIA RTX 2070 and the VRAM is 8G.
It's not enough to calculate the gradient of this function with NUM_VAR=10**7, but I really need it.

So I have two plans to do.
The first one is to run the code on better video cards with 16G VRAM.
The second one is to distribute the calculation.
But I find that the support for distributed calculation in jax is not so good, and it's now on experiencing step, right?

I think the second one is the better choice.. Can you give me some advice with this?

**Comment by mattjj** [comment]
Glad that helped!

Can you open a new issue for any additional JAX-specific questions?
